src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -611,9 +611,20 @@ private Tests MakeTestsAndBindingsForNegatedPattern(BoundDagTemp input, BoundNeg
             BoundRelationalPattern rel,
             out BoundDagTemp output)
         {
+            // check if the test is always true or always false
             var tests = ArrayBuilder<Tests>.GetInstance(2);
             output = MakeConvertToType(input, rel.Syntax, rel.Value.Type!, isExplicitTest: false, tests);
-            tests.Add(new Tests.One(new BoundDagRelationalTest(rel.Syntax, rel.Relation, rel.ConstantValue, output, rel.HasErrors)));
+            var fac = ValueSetFactory.ForSpecialType(input.Type.SpecialType);
+            var values = fac?.Related(rel.Relation.Operator(), rel.ConstantValue);
+            if (values?.IsEmpty == true)
+            {
+                tests.Add(Tests.False.Instance);
+            }
+            else if (values?.Complement().IsEmpty != true)
+            {
+                tests.Add(new Tests.One(new BoundDagRelationalTest(rel.Syntax, rel.Relation, rel.ConstantValue, output, rel.HasErrors)));
+            }
```

Comment:
Consider wrapping block in `if (fac is { }) { ... }` to avoid multiple `?.` operators and `== true`.

Reply:
I tried that, but ended up having to repeat line 625 in multiple else clauses.

---
In reply to: [395854215](https://github.com/dotnet/roslyn/pull/42501#discussion_r395854215) [](ancestors = 395854215)